### PR TITLE
cmake: support grouping of compile options for compiler testing

### LINKED ISSUE #47588: tests: sprintf: zero-length gnu_printf format string
**Describe the bug**
The print function is called and the function is not intended to be used to print characters, in this case shows build failure.

For example :
tests/lib/sprintf/src/main.c:778:24    ret = fprintf(stdout, "");
usb/device/usb_dc_kinetis.c:840:10     LOG_DBG("");
tests/lib/sprintf/src/main.c:760:15    ret = printf("");

It seems to be an error to print noting with twister command , but it is fine to use west build command.

Take tests/lib/sprintf/ as an example .

**To Reproduce**
1. source zephyr-env.sh
2. twister -p frdm_k64f --device-testing --hardware-map map.yml -T tests/lib/sprintf/ -vv
3. See error

**Expected behavior**
The test passes.

**Impact**
The test can not run by twister command.

**Logs and console output**
```
[25/136] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj
ccache /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBOARD_FLASH_SIZE="CONFIG_FLASH_SIZE*1024" -DCPU_MK64FN1M0VLL12 -DKERNEL -DTC_RUNID=ec44835519e9be9f92aef986773c3ea9 -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/ztest/zephyrproject/zephyr/include/zephyr -I/home/ztest/zephyrproject/zephyr/include -I/home/ztest/zephyrproject/zephyr/twister-out/frdm_k64f/tests/lib/sprintf/libraries.libc.sprintf/zephyr/include/generated -I/home/ztest/zephyrproject/zephyr/soc/arm/nxp_kinetis/k6x -I/home/ztest/zephyrproject/zephyr/soc/arm/nxp_kinetis/common -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/include -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/include/zephyr -I/home/ztest/zephyrproject/zephyr/subsys/testsuite/ztest/include/zephyr -I/home/ztest/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MK64F12 -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MK64F12/drivers -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/common -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/uart -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/port -I/home/ztest/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/sysmpu -isystem /home/ztest/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /home/ztest/zephyrproject/zephyr/twister-out/frdm_k64f/tests/lib/sprintf/libraries.libc.sprintf/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mfp16-format=ieee --sysroot=/home/ztest/zephyr-sdk-0.14.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/ztest/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/ztest/zephyrproject/zephyr/tests/lib/sprintf=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/ztest/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/ztest/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c: In function 'test_print':
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c:760:15: error: zero-length gnu_printf format string [-Werror=format-zero-length]
  760 |  ret = printf("");
      |               ^~
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c: In function 'test_fprintf':
/home/ztest/zephyrproject/zephyr/tests/lib/sprintf/src/main.c:778:24: error: zero-length gnu_printf format string [-Werror=format-zero-length]
  778 |  ret = fprintf(stdout, "");
      |                        ^~
cc1: all warnings being treated as errors
[26/136] Building C object zephyr/lib/posix/CMakeFiles/lib__posix.dir/pthread_common.c.obj
[27/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/abort.c.obj
[28/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoul.c.obj
[29/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoll.c.obj
[30/136] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/strtoull.c.obj
[31/136] Building C object zephyr/lib/posix/CMakeFiles/lib__posix.dir/nanosleep.c.obj
[32/136] Building C object zephyr/arch/arch/arm/core/aarch32/mpu/CMakeFiles/arch__arm__core__aarch32__mpu.dir/nxp_mpu.c.obj
[33/136] Building C object zephyr/subsys/testsuite/ztest/CMakeFiles/subsys__testsuite__ztest.dir/src/ztest.c.obj
[34/136] Building C object zephyr/arch/arch/arm/core/aarch32/cortex_m/CMakeFiles/arch__arm__core__aarch32__cortex_m.dir/fault.c.obj
ninja: build stopped: subcommand failed.

```

**Environment :**
OS: Linux,
Toolchain Zephyr SDK 0.14.1
Commit SHA or Version: b161554ee862

**Additional context**
The same error here: 
tests/lib/sprintf/src/main.c:760:15                  ret = printf("");
tests/lib/sprintf/src/main.c:778:24                  ret = fprintf(stdout, "");
drivers/usb/device/usb_dc_kinetis.c:188:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:188:2      LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:221:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:251:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:261:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:810:10    LOG_DBG("");
drivers/usb/device/usb_dc_kinetis.c:840:10    LOG_DBG("");
subsys/usb/device/class/hid/core.c:617:10     LOG_DBG("");
subsys/usb/device/class/hid/core.c:617:2       LOG_DBG("");

### PR DESCRIPTION
**Pull request by tejlmand**
Fixes: #47588

Kitware decided to fail compiler tests of single options if the
compiler reported only a warning and not an error:
https://github.com/Kitware/CMake/commit/f745e0497ee71d35fd1b3524b1636a72da76c266
which affects CMake 3.23.0 and 3.23.1

This change was later reverted in CMake >=3.23.2 and >=3.24.0
https://github.com/Kitware/CMake/commit/4941887d7defecb3016d2bd94d3a45754251ca56

Although the immediate issue is related to CMake, and has been fixed
reverted in later releases, then it still makes sense to be able to
group compiler options together that generally are depending on each
other to function correctly.

This commit introduces the possibility to group compiler options which
must be tested together.
It uses same approach as
```
add_compile_options("SHELL:<option1> <option2> ...")
```

Usage:
```
check_set_compiler_property(PROPERTY <property>
                            "SHELL:<option1> <option2> ..."
)
```

Will test the option together.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

**Comment by tejlmand** [comment]
although the initial error only exists in CMake 3.23.0 and 3.23.1 and got reverted for >=3.24.0 and >=3.23.2 then I believe it still make sense to have the ability to specify which options should be tested together.

**Comment by tejlmand** [comment]
@shaoanxli cannot add you as reviewer, but feel free to test it out.

File: cmake/modules/extensions.cmake [hunk 1]
```diff
@@ -2021,16 +2027,21 @@ function(check_set_compiler_property)
   list(REMOVE_AT COMPILER_PROPERTY_PROPERTY 0)
 
   foreach(option ${COMPILER_PROPERTY_PROPERTY})
+    if(${option} MATCHES "^SHELL:")
+      string(REGEX REPLACE "^SHELL:" "" option ${option})
+      separate_arguments(option UNIX_COMMAND ${option})
+    endif()
```
Comment (stephanosio) [inline cmake/modules/extensions.cmake:2033]: This feels like a hack to me, but I guess this is the best we can do with the limitations of the CMake language ...

**Review by stephanosio — APPROVED** [review]

File: cmake/modules/extensions.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2033]: Depends on how you look at it.
It's the same principle CMake itselfs uses, so if it is good enough to them, I assume we can use the same principle:

Ref: https://cmake.org/cmake/help/latest/command/add_compile_options.html#option-de-duplication
> One may specify a group of options using shell-like quoting along with a SHELL: prefix. The SHELL: prefix is dropped, and the rest of the option string is parsed using the [separate_arguments()](https://cmake.org/cmake/help/latest/command/separate_arguments.html#command:separate_arguments) UNIX_COMMAND mode.

**Review by nordicjm — APPROVED** [review]

**Pull request merged**